src/Authentication/Wsse.php

```diff
@@ -23,24 +23,33 @@ final class Wsse implements Authentication
     private $password;
 
     /**
-     * @param string $username
-     * @param string $password
+     * @var bool
      */
-    public function __construct($username, $password)
+    private $useSha512;
+
+    /**
+     * @param string     $username
+     * @param string     $password
+     * @param bool|false $useSha512
+     */
+    public function __construct($username, $password, $useSha512 = false)
     {
         $this->username = $username;
         $this->password = $password;
+        $this->useSha512 = $useSha512;
     }
 
     /**
      * {@inheritdoc}
      */
     public function authenticate(RequestInterface $request)
     {
-        // TODO: generate better nonce?
         $nonce = substr(md5(uniqid(uniqid().'_', true)), 0, 16);
         $created = date('c');
         $digest = base64_encode(sha1(base64_decode($nonce).$created.$this->password, true));
+        if (true === $this->useSha512) {
+            $digest = base64_encode(hash('sha512', base64_decode($nonce).$created.$this->password, true));
```

Comment:
as the argument to `hash` is a string, could we make the constructor argument generic to pass whatever hashing algorithm is desired by the user? the useSha512 flag will become cumbersome if some point in the future there is sha1024 or whatever. if we just pass a string, we have maximum flexibility. we can phpdoc the $hashAlgorithm parameter to say `To use a better hashing algorithm than the weak sha1, pass the algorithm to use, e.g. "sha512"`